lib/matplotlib/axes/_axes.py

```diff
@@ -2618,24 +2627,40 @@ def stem(self, *args, linefmt=None, markerfmt=None, basefmt=None, bottom=0,
         else:
             basestyle, basemarker, basecolor = _process_plot_format(basefmt)
 
+        # New behaviour in 3.1 is to use a LineCollection for the stemlines
+        if use_line_collection:
+            stemlines = []
+            for thisx, thisy in zip(x, y):
+                stemlines.append(((thisx, bottom), (thisx, thisy)))
+            stemlines = mcoll.LineCollection(stemlines, linestyles=linestyle,
+                                             colors=linecolor,
+                                             label='_nolegend_')
+            self.add_collection(stemlines)
+        # Old behaviour is to plot each of the lines individually
+        else:
+            cbook._warn_external(
+                'In Matplotlib 3.3 individual lines on a stem plot will be '
+                'added as a LineCollection instead of individual lines.\n'
+                'This significantly improves the performance of a stem plot.\n'
+                'To remove this warning and switch to the new behaviour, '
+                'set the "use_line_collection" keyword argument to True.')
+            stemlines = []
+            for thisx, thisy in zip(x, y):
+                l, = self.plot([thisx, thisx], [bottom, thisy],
+                               color=linecolor, linestyle=linestyle,
+                               marker=linemarker, label="_nolegend_")
+            stemlines.append(l)
```

Comment:
Indentation got lost. `append` must be within the loop. Or change to a list comprehension:

~~~
stemlines = [self.plot([xi, xi], [bottom, yi],
                       ...,
                       ...)[0]
             for xi, yi in zip(x, y)]
~~~